File: citadel/citadel-trading-comp/cpp/src/bot.cpp
#include "bot.hpp"
#include <iostream>
#include <chrono>
#include <thread>

namespace citadel {

TradingBot::TradingBot(std::shared_ptr<TradingApi> api) : api_(api) {
    exchange_arb_ = std::make_unique<ExchangeArbitrage>(api);
    index_arb_ = std::make_unique<IndexArbitrage>(api);
    shock_handler_ = std::make_unique<ShockHandler>(api);
}

bool TradingBot::run_once() {
    try {
        auto case_status = api_->get_case();
        if (case_status.status == "STOPPED") return false;

        if (case_status.tick != last_tick_) {
            last_tick_ = case_status.tick;
            std::cout << "Tick: " << last_tick_ << std::endl;

            shock_handler_->run_with_tick(last_tick_);
            exchange_arb_->run();
            index_arb_->run();
        }
    } catch (const std::exception& e) {
        std::cerr << "Bot error: " << e.what() << std::endl;
    }
    return true;
}

void TradingBot::start(double interval_seconds) {
    std::cout << "Starting C++ trading bot..." << std::endl;
    while (run_once()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(interval_seconds * 1000)));
    }
}

} // namespace citadel

File: citadel/citadel-trading-comp/rust/src/bot.rs
use crate::api::TradingApi;
use crate::strategies::{ExchangeArbitrage, IndexArbitrage, ShockHandler};
use std::thread;
use std::time::Duration;

pub struct TradingBot<'a> {
    api: &'a dyn TradingApi,
    last_tick: i32,
}

impl<'a> TradingBot<'a> {
    pub fn new(api: &'a dyn TradingApi) -> Self {
        Self { api, last_tick: -1 }
    }

    pub fn start(&mut self, interval_seconds: f64) {
        println!("Starting Rust trading bot...");
        while self.run_once() {
            thread::sleep(Duration::from_millis((interval_seconds * 1000.0) as u64));
        }
    }

    pub fn run_once(&mut self) -> bool {
        match self.api.get_case() {
            Ok(case) => {
                if case.status == "STOPPED" { return false; }

                if case.tick != self.last_tick {
                    self.last_tick = case.tick;
                    println!("Tick: {}", self.last_tick);

                    let shock_handler = ShockHandler::new(self.api);
                    let exchange_arb = ExchangeArbitrage::new(self.api);
                    let index_arb = IndexArbitrage::new(self.api);

                    let _ = shock_handler.run(case.tick);
                    exchange_arb.run();
                    let _ = index_arb.run();
                }
                true
            }
            Err(e) => {
                eprintln!("Bot error: {}", e);
                true
            }
        }
    }
}
